### PR TITLE
Skip LayerTree::Preroll LayerTree::Paint & Swapbuffer when frame_damage is empty

### PR DESCRIPTION
**Pull request by linxuebin1990**
See https://github.com/flutter/flutter/issues/101097

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by zanderso** [comment]
From PR review triage: This PR will need tests.

cc @knopp

**Comment by flar** [comment]
I double checked that we don't end up calling CleanupAfterFrame here which would empty the raster cache, so that shouldn't be an issue.

I don't see any problems, but I would rather @iskakaushik takes a look at this to double-check.

**Review by iskakaushik — APPROVED** [review]
LGTM. Nice fix!

**Comment by linxuebin1990** [comment]
> From PR review triage: This PR will need tests.
> 
Hi @zanderso 
Thanks for the reply, I have already add the test code, do you have any other suggestions?

**Comment by linxuebin1990** [comment]
> I double checked that we don't end up calling CleanupAfterFrame here which would empty the raster cache, so that shouldn't be an issue.
> 
> I don't see any problems, but I would rather @iskakaushik takes a look at this to double-check.

Hi @flar
Thanks for the code review! I have already add the test code, do you have any other suggestions?

**Review by flar — APPROVED** [review]
echo the LGTM from @iskakaushik

File: shell/common/rasterizer_unittests.cc [hunk 1]
```diff
@@ -726,4 +729,139 @@ TEST(
...
+                                 false,                   // ignore raster cache
+                                 last_damage.get()        // frame damage
+        );
+    EXPECT_EQ(last_raster_status, RasterStatus::kSuccess);
+
+    // Draw Current Frame
+    std::unique_ptr<FrameDamage> cur_damage = std::make_unique<FrameDamage>();
+    cur_damage->SetPreviousLayerTree(last_layer_tree.get());
+    cur_damage->AddAdditonalDamage(SkIRect());
+    ;
```
Comment (flutter-hacker) [inline shell/common/rasterizer_unittests.cc:789]: ;

**Pull request merged**